Tribler/Policies/BoostingSource.py

```diff
@@ -0,0 +1,559 @@
+from Tribler.community.allchannel.community import AllChannelCommunity
+from Tribler.community.channel.community import ChannelCommunity
+from Tribler.dispersy.exception import CommunityNotFoundException
+from Tribler.dispersy.util import blocking_call_on_reactor_thread
+
+
+class BoostingSource(object):
+    """
+    Base class for boosting source. For now, it can be RSS, directory, and channel
+    """
+
+    def __init__(self, session, source, boost_settings, callback):
+        self.session = session
+        self.channelcast_db = session.lm.channelcast_db
+
+        self.torrents = {}
+        self.source = source
+        self.interval = boost_settings.source_interval
+        self.max_torrents = boost_settings.max_torrents_per_source
+        self.callback = callback
+        self.archive = False
+
+        self.enabled = True
+
+        self.av_uprate = 0
+        self.av_dwnrate = 0
+        self.storage_used = 0
+        self.ready = False
+
+        self.min_connection = boost_settings.min_connection_start
+        self.min_channels = boost_settings.min_channels_start
+
+        self.torrent_mgr = TorrentManagerCM.get_instance(session)
+
+        # local import for handling circular import
+        from Tribler.Policies.BoostingManager import BoostingManager
+        self.boosting_manager = BoostingManager.get_instance()
+        self._logger = logging.getLogger(BoostingManager.__name__)
+
+    def kill_tasks(self):
+        """
+        kill tasks on this source
+        """
+
+        self.ready = False
+        self.torrent_mgr.del_instance()
+
+    def _load_if_ready(self, source):
+        """
+        load source if and only if the overall system is ready.
+
+        Useful to not burden the apps in startup
+        """
+        def check_system(defer_param=None):
+            """
+            function that check the system whether it's ready or not
+
+            it depends on #connection and #channel
+            """
+            if defer_param is None:
+                defer_param = defer.Deferred()
+
+            nr_channels = self.channelcast_db.getNrChannels()
+            nr_connections = 0
+
+            for community in self.session.lm.dispersy.get_communities():
+                from Tribler.community.search.community import SearchCommunity
+                if isinstance(community, SearchCommunity):
+                    nr_connections = community.get_nr_connections()
+
+            if nr_channels > self.min_channels and nr_connections > self.min_connection:
+                defer_param.callback(source)
+            else:
+                self.boosting_manager.register_task(str(self.source)+"_check_sys",
+                                                    reactor.callLater(10, check_system, defer_param))
+
+
+            return defer_param
+
+        defer_check = check_system()
+        defer_check.addCallback(self._load)
+        defer_check.addErrback(self._on_err)
```

Comment:
I see. I thought it will be nice to call it in different function because it more readable for me.